File: mobile/lib/utils/id_card_validator.dart
/// 身份证号验证工具类
class IdCardValidator {
  // 地区码映射表（部分）
  static const Map<String, String> _areaCodes = {
    '11': '北京市',
    '12': '天津市',
    '13': '河北省',
    '14': '山西省',
    '15': '内蒙古自治区',
    '21': '辽宁省',
    '22': '吉林省',
    '23': '黑龙江省',
    '31': '上海市',
    '32': '江苏省',
    '33': '浙江省',
    '34': '安徽省',
    '35': '福建省',
    '36': '江西省',
    '37': '山东省',
    '41': '河南省',
    '42': '湖北省',
    '43': '湖南省',
    '44': '广东省',
    '45': '广西壮族自治区',
    '46': '海南省',
    '50': '重庆市',
    '51': '四川省',
    '52': '贵州省',
    '53': '云南省',
    '54': '西藏自治区',
    '61': '陕西省',
    '62': '甘肃省',
    '63': '青海省',
    '64': '宁夏回族自治区',
    '65': '新疆维吾尔自治区',
    '71': '台湾省',
    '81': '香港特别行政区',
    '82': '澳门特别行政区',
  };

  // 权重因子
  static const List<int> _weightFactors = [
    7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2
  ];

  // 校验码映射
  static const List<String> _checkCodes = [
    '1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2'
  ];

  /// 验证身份证号是否合法
  static ValidationResult validate(String idCard) {
    if (idCard.isEmpty) {
      return ValidationResult(
        isValid: false,
        error: '身份证号不能为空',
      );
    }

    // 去除空格
    idCard = idCard.trim();

    // 长度检查
    if (idCard.length != 18) {
      return ValidationResult(
        isValid: false,
        error: '身份证号必须为18位',
      );
    }

    // 格式检查：前17位为数字，第18位为数字或X
    final regex = RegExp(r'^\d{17}[\dXx]$');
    if (!regex.hasMatch(idCard)) {
      return ValidationResult(
        isValid: false,
        error: '身份证号格式不正确',
      );
    }

    // 地区码检查
    final areaCode = idCard.substring(0, 2);
    if (!_areaCodes.containsKey(areaCode)) {
      return ValidationResult(
        isValid: false,
        error: '身份证号地区码无效',
      );
    }

    // 出生日期检查
    final birthDateStr = idCard.substring(6, 14);
    if (!_isValidDate(birthDateStr)) {
      return ValidationResult(
        isValid: false,
        error: '身份证号出生日期无效',
      );
    }

    // 校验码检查
    final checkCode = _calculateCheckCode(idCard);
    final lastChar = idCard[17].toUpperCase();
    if (lastChar != checkCode) {
      return ValidationResult(
        isValid: false,
        error: '身份证号校验码错误',
      );
    }

    return ValidationResult(
      isValid: true,
      area: _areaCodes[areaCode],
      birthDate: _formatBirthDate(birthDateStr),
    );
  }

  /// 计算校验码
  static String _calculateCheckCode(String idCard) {
    int sum = 0;
    for (int i = 0; i < 17; i++) {
      sum += int.parse(idCard[i]) * _weightFactors[i];
    }
    return _checkCodes[sum % 11];
  }

  /// 验证日期是否有效
  static bool _isValidDate(String dateStr) {
    try {
      final year = int.parse(dateStr.substring(0, 4));
      final month = int.parse(dateStr.substring(4, 6));
      final day = int.parse(dateStr.substring(6, 8));

      if (year < 1900 || year > DateTime.now().year) return false;
      if (month < 1 || month > 12) return false;

      final daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
      // 闰年判断
      if ((year % 4 == 0 && year % 100 != 0) || year % 400 == 0) {
        daysInMonth[1] = 29;
      }

      if (day < 1 || day > daysInMonth[month - 1]) return false;

      return true;
    } catch (e) {
      return false;
    }
  }

  /// 格式化出生日期
  static String _formatBirthDate(String dateStr) {
    return '${dateStr.substring(0, 4)}-${dateStr.substring(4, 6)}-${dateStr.substring(6, 8)}';
  }

  /// 脱敏身份证号（只显示前3位和后4位）
  static String mask(String idCard) {
    if (idCard.length != 18) return idCard;
    return '${idCard.substring(0, 3)}***********${idCard.substring(14)}';
  }

  /// 获取年龄
  static int? getAge(String idCard) {
    final result = validate(idCard);
    if (!result.isValid || result.birthDate == null) return null;

    final birthDate = DateTime.tryParse(result.birthDate!);
    if (birthDate == null) return null;

    final now = DateTime.now();
    int age = now.year - birthDate.year;
    if (now.month < birthDate.month ||
        (now.month == birthDate.month && now.day < birthDate.day)) {
      age--;
    }
    return age;
  }

  /// 获取性别（0=女，1=男）
  static int? getGender(String idCard) {
    if (idCard.length != 18) return null;
    final genderDigit = int.tryParse(idCard[16]);
    if (genderDigit == null) return null;
    return genderDigit % 2; // 奇数为男，偶数为女
  }
}

class ValidationResult {
  final bool isValid;
  final String? error;
  final String? area;
  final String? birthDate;

  ValidationResult({
    required this.isValid,
    this.error,
    this.area,
    this.birthDate,
  });
}

File: deploy_package/src/main/java/com/lanmouse/util/IdCardValidator.java
package com.lanmouse.util;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;

/**
 * 身份证号工具类
 */
public class IdCardValidator {

    // 权重值
    private static final int[] WEIGHT = {7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2};

    // 校验码
    private static final char[] CHECK_CODE = {'1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2'};

    // 有效的省份代码（简化版）
    private static final String[] PROVINCE_CODES = {
        "11", "12", "13", "14", "15",
        "21", "22", "23",
        "31", "32", "33", "34", "35", "36", "37",
        "41", "42", "43", "44", "45", "46",
        "50", "51", "52", "53", "54",
        "61", "62", "63", "64", "65",
        "71", "81", "82"
    };

    /**
     * 验证身份证号合法性
     *
     * @param idCard 身份证号
     * @return 是否合法
     */
    public static boolean isValid(String idCard) {
        if (idCard == null || idCard.length() != 18) {
            return false;
        }

        // 格式验证
        if (!idCard.matches("^\\d{17}[\\dXx]$")) {
            return false;
        }

        // 省份验证
        String provinceCode = idCard.substring(0, 2);
        if (!isValidProvinceCode(provinceCode)) {
            return false;
        }

        // 出生日期验证
        String birthDateStr = idCard.substring(6, 14);
        if (!isValidBirthDate(birthDateStr)) {
            return false;
        }

        // 校验码验证
        return validateCheckCode(idCard);
    }

    /**
     * 验证省份代码
     */
    private static boolean isValidProvinceCode(String code) {
        for (String validCode : PROVINCE_CODES) {
            if (validCode.equals(code)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 验证出生日期
     */
    private static boolean isValidBirthDate(String dateStr) {
        try {
            LocalDate birthDate = LocalDate.parse(dateStr, DateTimeFormatter.ofPattern("yyyyMMdd"));

            // 检查日期是否在过去和未来合理范围内
            LocalDate now = LocalDate.now();
            if (birthDate.isAfter(now) || birthDate.isBefore(now.minusYears(150))) {
                return false;
            }

            return true;
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    /**
     * 验证校验码
     */
    private static boolean validateCheckCode(String idCard) {
        int sum = 0;
        for (int i = 0; i < 17; i++) {
            sum += Character.getNumericValue(idCard.charAt(i)) * WEIGHT[i];
        }

        int index = sum % 11;
        char expectedCheckCode = CHECK_CODE[index];
        char actualCheckCode = Character.toUpperCase(idCard.charAt(17));

        return expectedCheckCode == actualCheckCode;
    }

    /**
     * 从身份证号提取出生日期
     *
     * @param idCard 身份证号
     * @return 出生日期 (yyyy-MM-dd)
     */
    public static String getBirthDate(String idCard) {
        if (!isValid(idCard)) {
            return null;
        }
        String dateStr = idCard.substring(6, 14);
        return dateStr.substring(0, 4) + "-" + dateStr.substring(4, 6) + "-" + dateStr.substring(6, 8);
    }

    /**
     * 从身份证号提取性别
     *
     * @param idCard 身份证号
     * @return "M" (男) 或 "F" (女)
     */
    public static String getGender(String idCard) {
        if (idCard.length() != 18) {
            return null;
        }
        int genderDigit = Character.getNumericValue(idCard.charAt(16));
        return (genderDigit % 2 == 0) ? "F" : "M";
    }

    /**
     * 从身份证号提取年龄
     *
     * @param idCard 身份证号
     * @return 年龄
     */
    public static int getAge(String idCard) {
        String birthDateStr = idCard.substring(6, 14);
        try {
            LocalDate birthDate = LocalDate.parse(birthDateStr, DateTimeFormatter.ofPattern("yyyyMMdd"));
            LocalDate now = LocalDate.now();
            int age = now.getYear() - birthDate.getYear();
            if (now.getDayOfYear() < birthDate.getDayOfYear()) {
                age--;
            }
            return age;
        } catch (DateTimeParseException e) {
            return 0;
        }
    }

    /**
     * 掩码身份证号（脱敏）
     *
     * @param idCard 身份证号
     * @return 掩码后的身份证号
     */
    public static String mask(String idCard) {
        if (idCard == null || idCard.length() < 8) {
            return "**************";
        }
        return idCard.substring(0, 3) + "********" + idCard.substring(idCard.length() - 4);
    }

    public static String maskIdCard(String idCard) {
        return mask(idCard);
    }

    public static String hashIdCard(String idCard) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest((idCard + "lanmouse_salt").getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 algorithm not found", e);
        }
    }

    /**
     * 获取身份证归属地
     *
     * @param idCard 身份证号
     * @return 归属地（简化版）
     */
    public static String getProvince(String idCard) {
        if (idCard == null || idCard.length() < 2) {
            return "未知";
        }

        String provinceCode = idCard.substring(0, 2);
        return getProvinceName(provinceCode);
    }

    private static String getProvinceName(String code) {
        switch (code) {
            case "11": return "北京市";
            case "12": return "天津市";
            case "13": return "河北省";
            case "14": return "山西省";
            case "15": return "内蒙古";
            case "21": return "辽宁省";
            case "22": return "吉林省";
            case "23": return "黑龙江省";
            case "31": return "上海市";
            case "32": return "江苏省";
            case "33": return "浙江省";
            case "34": return "安徽省";
            case "35": return "福建省";
            case "36": return "江西省";
            case "37": return "山东省";
            case "41": return "河南省";
            case "42": return "湖北省";
            case "43": return "湖南省";
            case "44": return "广东省";
            case "45": return "广西";
            case "46": return "海南省";
            case "50": return "重庆市";
            case "51": return "四川省";
            case "52": return "贵州省";
            case "53": return "云南省";
            case "54": return "西藏";
            case "61": return "陕西省";
            case "62": return "甘肃省";
            case "63": return "青海省";
            case "64": return "宁夏";
            case "65": return "新疆";
            case "71": return "台湾";
            case "81": return "香港";
            case "82": return "澳门";
            default: return "未知";
        }
    }
}
